R/calcGAINS2025forREMIND.R

```diff
@@ -0,0 +1,177 @@
+#' Calculates air pollutant emissions, activities and emission factors
+#' for all scenarios and SSPs available from GAINS, including both
+#' legacy and 2025 variants of GAINS data. This generates the actual
+#' files needed for REMIND in the proper format, for the scenario logic,
+#' see `GAINS2025scenarios`.
+#'
+#' @return Activity levels, emissions or emission factors
+#' @author Gabriel Abrahao
+#' @param subtype "emission_factors", "emissions","emissions_starting_values"
+#'
+#' @importFrom abind abind
+#' @importFrom utils head tail
+#' @importFrom magclass as.magpie
+#' @importFrom tidyr pivot_longer drop_na
+calcGAINS2025forREMIND <- function(subtype) {
+  # Binds new and old versions of GAINS data, adding an "ssp" dimension
+  # to old data and makes the new data conform to the old shape besides the
+  # new dimension
+  bindNewOld <- function(innew, inold) {
+    # Assume a dummy "GAINSlegacy" SSP for the old data to fill the dimension
+    newlastdim <- 3 + (as.integer(substr(tail(names(getSets(inold)), 1), 4, 4)) + 1) * 1e-1
+    old <- toolAddDimensions(inold, "GAINSlegacy", "ssp", newlastdim)
+    getSets(innew) <- c("region", "year", "scenario", "ssp", "sector", "emi")
+    outdimorder <- match(getSets(innew), getSets(old))[3:length(getSets(innew))] - 2
+    innew <- dimOrder(innew, outdimorder)
+
+    # Sectors absent in new data, should all have zero in the old data
+    # Pad new data with these old sectors
+    wstsecs <- setdiff(getItems(old, "sector"), getItems(innew, "sector"))
+    dummy <- old[, , wstsecs][, , "SSP2"]
+    if (!all(dummy == 0)) {
+      warning(paste0(
+        "When padding sectors in GAINS2025 with those from GAINSlegacy, found non-zero values in sectors: ",
+        paste(wstsecs, collapse = " ")
+      ))
+    }
+    dummy <- mbind(lapply(getItems(innew, "scenario"), \(x) setItems(dummy, "scenario", x)))
+    dummy <- mbind(lapply(getItems(innew, "ssp"), \(x) setItems(dummy, "ssp", x)))
+
+    out <- mbind(old, innew, dummy)
+    getComment(out)[1] <- getFromComment(innew, "description")
+    getComment(out)[2] <- getFromComment(innew, "unit")
+    return(out)
+  }
+
+
+  if (subtype == "emissions") {
+    inold <- calcOutput("GAINSEmi", subtype = "emissions", aggregate = F)
+    innew <- calcOutput("GAINS2025scenarios", subtype = "emissions", aggregate = F)
+    out <- bindNewOld(innew, inold)
+
+    wgt <- NULL
+    desc <- getFromComment(innew, "description")
+    unit <- getFromComment(innew, "unit")
+  } else if (subtype == "emission_factors") {
+    linold <- calcOutput("GAINSEmi", subtype = "emission_factors", aggregate = F, supplementary = T)
+    inold <- linold$x
+    linnew <- calcOutput("GAINS2025scenarios", subtype = "emission_factors", aggregate = F, supplementary = T)
+    innew <- linnew$x
+    out <- bindNewOld(innew, inold)
+
+    # Weights, GAINS2025 uses activities that are per SSP but not per scenario, pad the dimensions
+    wgt <- bindNewOld(linnew$w, linold$w)
```

Comment:
The chosen weight causes two warnings

```
1: In toolAggregate(x = x$x, weight = x$weight, rel = map$rel) :
  Negative numbers in weight. Dangerous, was it really intended?
2: In toolAggregate(x = x$x, weight = x$weight, rel = map$rel) :
  Weight sum is 0, so cannot normalize and will return 0 for some aggregation targets. This changes the total sum of the magpie object! If this is really intended set zeroWeight = "allow", or "setNA" to return NA.
  ```

Is this really intended, or are these actual problems?

Reply:
The negative weights are really weird, as that's supposed to be activity data, but they are present in the original data for some reason in OCEA_WHOL (Oceania) and TURK_WHOL (Turkey) for `Transformations_HLF` and `Transformations_NatGas`. I set those negative values to 0 for now, noting that emissions are also zero in those cases.

Regarding the zeros, I think that's intended, please correct me if my reasoning is wrong. We are aggregating emission factors (emissions per quantitiy of activity, say, tons of SO2 per PJ of energy generated). If the activity in a certain region is zero, it shouldn't count towards the aggregated emission factors. Those are supposed to have emission factors = 0 anyway, only there because NAs are a pain.

Reply:
> please correct me if my reasoning is wrong

Sounds logical to me, so 'zeroWeight = "allow"` should be the way to go.